libkmod/libkmod-index.c

```diff
@@ -833,10 +833,10 @@ static void index_mm_dump_node(struct index_mm_node *node, struct strbuf *buf, i
 		struct index_mm_value v;
 
 		read_value_mm(&p, &v);
-		write_str_safe(fd, buf->bytes, buf->used);
-		write_str_safe(fd, " ", 1);
-		write_str_safe(fd, v.value, v.len);
-		write_str_safe(fd, "\n", 1);
+		fwrite(buf->bytes, 1, buf->used, fp);
+		fputc(' ', fp);
+		fwrite(v.value, 1, v.len, fp);
+		fputc('\n', fp);
```

Comment:
Out of curiosity: have you tried placing the contents into a buffer (on the stack) and doing a single write for given value instead of 4? Does it make a difference?

Reply:
This already helps by almost 50 %, which is nice. At first I tried scratchbuf (and soon strbuf), but we could also keep a buffer around for longer. See https://github.com/kmod-project/kmod/pull/252 as an example.